File: TheaterSuggestions/CSharp/ScreeningSeatingArrangementSuggestions/DeepModeling/AdjacentSeatingPlace/AdjacentSeatingPlaces.cs
namespace SeatsSuggestions.DeepModeling.AdjacentSeatingPlace;

public static class AdjacentSeatingPlaces
{
    public static IEnumerable<SeatingPlace> OfferAdjacentSeatingPlace(
        IEnumerable<SeatingPlaceWithDistance> seatingPlacesWithDistances, PricingCategory pricingCategory, int partySize)
    {
        var potentialAdjacentSeats = new List<SeatingPlaceWithDistance>();
        var groupsOfAdjacentSeats = new List<GroupOfAdjacentSeats>();
        var seatingPlacesWithDistancePrevious =
            new SeatingPlaceWithDistance(
                new SeatingPlace("Z", -1, PricingCategory.Ignored, SeatingPlaceAvailability.Allocated), -1);

        foreach (var seatingPlaceWithDistance in seatingPlacesWithDistances
                     .OrderBy(seatingPlaceWithDistance => seatingPlaceWithDistance.SeatingPlace.Number)
                     .Where(s => s.SeatingPlace.IsAvailable())
                     .Where(s => s.SeatingPlace.MatchCategory(pricingCategory))
                 )
        {
            if (seatingPlacesWithDistancePrevious.AreAdjacentSeats(seatingPlaceWithDistance))
            {
                if (!potentialAdjacentSeats.Any())
                {
                    potentialAdjacentSeats.Add(seatingPlacesWithDistancePrevious);
                }

                potentialAdjacentSeats.Add(seatingPlaceWithDistance);
            }
            else
            {
                potentialAdjacentSeats.Clear();
            }

            seatingPlacesWithDistancePrevious = seatingPlaceWithDistance;

            if (partySize.Matches(potentialAdjacentSeats))
            {
                seatingPlacesWithDistancePrevious =
                    groupsOfAdjacentSeats.AddGroupOfPlaceWithDistances(potentialAdjacentSeats);
            }
        }

        return groupsOfAdjacentSeats.Any()
            ? SelectTheBestGroupOfSeatingPlaces(groupsOfAdjacentSeats)
            : new List<SeatingPlace>();
    }

    private static IEnumerable<SeatingPlace> SelectTheBestGroupOfSeatingPlaces(
        List<GroupOfAdjacentSeats> groupsOfAdjacentSeats)
    {
        var bestOfGroupOfAdjacentSeats = groupsOfAdjacentSeats.OrderBy(s => s.SumOfDistance).First();
        return bestOfGroupOfAdjacentSeats.SeatingPlaces.OrderBy(seatingPlace => seatingPlace.Number);
    }
}
internal class GroupOfAdjacentSeats(List<SeatingPlace> seatingPlaces, int sumOfDistance)
{
    public List<SeatingPlace> SeatingPlaces { get; } = seatingPlaces;
    public int SumOfDistance { get; } = sumOfDistance;
}

File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestionsDomain/src/main/java/org/octo/seatingplacessuggestions/seatingplacesuggestions/deepmodel/adjacentseatingplace/AdjacentSeatingPlaces.java
package org.octo.seatingplacessuggestions.seatingplacesuggestions.deepmodel.adjacentseatingplace;

import org.octo.seatingplacessuggestions.seatingplacesuggestions.PricingCategory;
import org.octo.seatingplacessuggestions.seatingplacesuggestions.SeatingPlace;
import org.octo.seatingplacessuggestions.seatingplacesuggestions.SeatingPlaceAvailability;
import org.octo.seatingplacessuggestions.seatingplacesuggestions.deepmodel.SeatingPlaceWithDistance;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class AdjacentSeatingPlaces {
    public static List<SeatingPlace> offerAdjacentSeatingPlace(List<SeatingPlaceWithDistance> seatingPlacesWithDistance, PricingCategory pricingCategory, int partySize)
    {
        var potentialAdjacentSeats = new ArrayList<SeatingPlaceWithDistance>();
        var groupsOfAdjacentSeats = new ArrayList<GroupOfAdjacentSeats>();
        var seatingPlacesWithDistancePrevious =
                new SeatingPlaceWithDistance(
                        new SeatingPlace("Z", -1, PricingCategory.IGNORED, SeatingPlaceAvailability.ALLOCATED), -1);

        List<SeatingPlaceWithDistance> seatingPlaceWithDistances = seatingPlacesWithDistance.stream()
                .filter(s -> s.SeatingPlace().matchCategory(pricingCategory))
                .filter(s -> s.SeatingPlace().isAvailable())
                .sorted(Comparator.comparing(s -> s.SeatingPlace().number())).toList();

        for(var seatingPlaceWithDistance : seatingPlaceWithDistances)
        {
            if (areAdjacentSeats(seatingPlacesWithDistancePrevious, seatingPlaceWithDistance))
            {
                if (potentialAdjacentSeats.isEmpty())
                {
                    potentialAdjacentSeats.add(seatingPlacesWithDistancePrevious);
                }

                potentialAdjacentSeats.add(seatingPlaceWithDistance);
            }
            else
            {
                potentialAdjacentSeats.clear();
            }

            seatingPlacesWithDistancePrevious = seatingPlaceWithDistance;

            if (partySize == potentialAdjacentSeats.size())
            {
                seatingPlacesWithDistancePrevious = addGroupOfPlaceWithDistances(groupsOfAdjacentSeats, potentialAdjacentSeats);
            }
        }

        return !groupsOfAdjacentSeats.isEmpty()
                ? selectTheBestGroupOfSeatingPlaces(groupsOfAdjacentSeats)
                : new ArrayList<>();

    }


    public static SeatingPlaceWithDistance addGroupOfPlaceWithDistances(
            List<GroupOfAdjacentSeats> groupsOfAdjacentSeats, List<SeatingPlaceWithDistance> potentialAdjacentSeats)
    {
        var sumOfDistance = (Integer)potentialAdjacentSeats.stream().mapToInt(SeatingPlaceWithDistance::DistanceFromTheMiddleOfTheRow).sum();
        var seatingPlaces = potentialAdjacentSeats.stream().map(SeatingPlaceWithDistance::SeatingPlace);
        groupsOfAdjacentSeats.add(new GroupOfAdjacentSeats(seatingPlaces.toList(), sumOfDistance));
        potentialAdjacentSeats.clear();
        return new SeatingPlaceWithDistance(
                new SeatingPlace("Z", -1, PricingCategory.IGNORED, SeatingPlaceAvailability.ALLOCATED), -1);
    }

    public static Boolean areAdjacentSeats(SeatingPlaceWithDistance seatingPlacesWithDistancePrevious,
                                           SeatingPlaceWithDistance seatingPlaceWithDistance)
    {
        return seatingPlacesWithDistancePrevious.SeatingPlace().number() + 1 ==
                seatingPlaceWithDistance.SeatingPlace().number();
    }

    private static List<SeatingPlace> selectTheBestGroupOfSeatingPlaces(
            List<GroupOfAdjacentSeats> groupsOfAdjacentSeats)
    {
        var bestOfGroupOfAdjacentSeats = groupsOfAdjacentSeats.stream()
                .sorted(Comparator.comparing(GroupOfAdjacentSeats::SumOfDistance))
                .toList();

        return bestOfGroupOfAdjacentSeats.get(0).SeatingPlaces();
    }
}
